service-base/src/main/java/org/eclipse/hono/service/credentials/validators/AbstractSecretsValidator.java

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright (c) 2017 Bosch Software Innovations GmbH.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial creation
+ */
+package org.eclipse.hono.service.credentials.validators;
+
+import org.eclipse.hono.service.credentials.SecretsValidator;
+import org.eclipse.hono.util.CredentialsConstants;
+import org.eclipse.hono.util.CredentialsObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+/**
+ * Abstract validator class from which concrete validators are derived that are specialized to a specific type of
+ * credential secrets.
+ * <p>
+ * The class implements the validation steps that are common to all types of credentials. This e.g. includes the enabled
+ * flag of credential entries, the iteration over several valid credential entries (until one is successfully validated
+ * or none is left), etc.
+ * <p>
+ * The detailed algorithm to validate a single credential entry is delegated to the implementing subclass and so supports
+ * a specified small implementation class per credentials type..
+ *
+ * @param <T> The type of what has to be validated (called item below): this can be String in case of password validation, a certificate
+ *           class in case of a client certificate, etc.
+ */
+public abstract class AbstractSecretsValidator<T> implements SecretsValidator<T> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractSecretsValidator.class);
+
+    /**
+     * Get the type of credentials secrets this validator is responsible for.
+     * <p>This can be freely defined, but there are some predefined types in the
+     * <a href="https://www.eclipse.org/hono/api/Credentials-API/">Credentials API</a>.
+     *
+     * @return The type of credentials secrets.
+     */
+    @Override
+    public abstract String getSecretsType();
```

Comment:
no need to re-define it here (the interface already defines this method as abstract)